core/server/worker/src/main/java/alluxio/worker/block/BlockLockManager.java

```diff
@@ -99,31 +99,60 @@ public BlockLockManager() {}
-      if (sessionHoldsLock(sessionId, blockId)) {
-        throw new IllegalStateException(String
-            .format("Session %s attempted to take a write lock on block %s, but the session already"
-                + " holds a lock on the block", sessionId, blockId));
-      }
-      lock = blockLock.writeLock();
+    Lock lock = blockLockType == BlockLockType.READ ? blockLock.readLock() : blockLock.writeLock();
+    // Make sure the session isn't already holding the block lock.
+    if (blockLockType == BlockLockType.WRITE && sessionHoldsLock(sessionId, blockId)) {
+      throw new IllegalStateException(String
+          .format("Session %s attempted to take a write lock on block %s, but the session already"
+              + " holds a lock on the block", sessionId, blockId));
     }
     lock.lock();
     try {
-      long lockId = LOCK_ID_GEN.getAndIncrement();
-      try (LockResource r = new LockResource(mSharedMapsLock.writeLock())) {
-        mLockIdToRecordMap.put(lockId, new LockRecord(sessionId, blockId, lock));
-        Set<Long> sessionLockIds = mSessionIdToLockIdsMap.get(sessionId);
-        if (sessionLockIds == null) {
-          mSessionIdToLockIdsMap.put(sessionId, Sets.newHashSet(lockId));
-        } else {
-          sessionLockIds.add(lockId);
-        }
+      return allocateLockId(sessionId, blockId, lock);
+    } catch (Throwable e) {
+      // If an unexpected exception occurs, we should release the lock to be conservative.
+      unlock(lock, blockId);
+      throw e;
+    }
+  }
+
+  /**
+   * Tries to lock a block within the given time.
+   * Note that even if this block does not exist, a lock id is still returned.
+   *
+   * If all {@link PropertyKey#WORKER_TIERED_STORE_BLOCK_LOCKS} are already in use and no lock has
+   * been allocated for the specified block, this method will need to wait until a lock can be
+   * acquired from the lock pool.
+   *
+   * @param sessionId the session id
+   * @param blockId the block id
+   * @param blockLockType {@link BlockLockType#READ} or {@link BlockLockType#WRITE}
+   * @param time the maximum time to wait for the lock
+   * @param unit the time unit of the {@code time} argument
+   * @return lock id or INVALID_LOCK_ID if not able to lock within the given time
+   */
+  public long tryLockBlock(long sessionId, long blockId, BlockLockType blockLockType,
+      long time, TimeUnit unit) {
+    ClientRWLock blockLock = getBlockLock(blockId);
+    Lock lock = blockLockType == BlockLockType.READ ? blockLock.readLock() : blockLock.writeLock();
+    // Make sure the session isn't already holding the block lock.
+    if (blockLockType == BlockLockType.WRITE && sessionHoldsLock(sessionId, blockId)) {
+      throw new IllegalStateException(String
+          .format("Session %s attempted to take a write lock on block %s, but the session already"
+              + " holds a lock on the block", sessionId, blockId));
+    }
+    try {
```

Comment:
in case `useTryLock=false`, `lockBlock(long, long, BlockLockType, boolean useTryLock, long, unit)`  requires some used time unit value. This is not a good way for an API. I would avoid convoluting two methods

Reply:
@apc999 Sorry for not clarify the detail, I mean create a `private` method `private void lockBlock(long, long, BlockLockType, boolean useTryLock, long, unit)`, so it's not an API, you can keep the two method `lockBlock` and `tryLockBlock` as the public method to be API for outer callee.

Reply:
The full changes I suggested is 
```java


  public long lockBlock(long sessionId, long blockId, BlockLockType blockLockType) {
    return internalLockBlock(sessionId, blockId, blockLockType, false, -1, null);
  }

  public long tryLockBlock(long sessionId, long blockId, BlockLockType blockLockType,
      long time, TimeUnit unit) {
    return internalLockBlock(sessionId, blockId, blockLockType, true, time, unit);
  }

  private long internalLockBlock(long sessionId, long blockId, BlockLockType blockLockType,
      boolean useTryLock, long time, TimeUnit unit) {
    ClientRWLock blockLock = getBlockLock(blockId);
    Lock lock = blockLockType == BlockLockType.READ ? blockLock.readLock() : blockLock.writeLock();
    // Make sure the session isn't already holding the block lock.
    if (blockLockType == BlockLockType.WRITE && sessionHoldsLock(sessionId, blockId)) {
      throw new IllegalStateException(String
          .format("Session %s attempted to take a write lock on block %s, but the session already"
              + " holds a lock on the block", sessionId, blockId));
    }
    if (useTryLock) {
      try {
        if (!lock.tryLock(time, unit)) {
          LOG.warn("Failed to acquire lock for block {} after {} {}.  "
                  + "session: {}, blockLockType: {}, lock reference count = {}",
              blockId, time, unit, sessionId, blockLockType,
              blockLock.getReferenceCount());
          return INVALID_LOCK_ID;
        }
      } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
      }
    } else {
      lock.lock();
    }
    try {
      return allocateLockId(sessionId, blockId, lock);
    } catch (Throwable e) {
      // If an unexpected exception occurs, we should release the lock to be conservative.
      unlock(lock, blockId);
      throw e;
    }
  }
```

Reply:
good call. I adopted what you suggested here. Thanks ! @maobaolong